### ISSUE TITLE
conch: support hmac-sha2-256 and hmac-sha2-512

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8108]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)| @cjwatson reported|
|-|-|
|Trac ID|trac#8108|
|Type|enhancement|
|Created|2015-11-16 00:54:50Z|
|Branch|https://github.com/twisted/twisted/tree/conch-ssh-hmac-sha2-8108|

RFC 6668 defines hmac-sha2-256 and hmac-sha2-512 MAC algorithms for SSH.  conch can easily support both, and should, since hmac-sha1 is beginning to look a bit creaky.

I'll attach a patch shortly to add these.

Attachments:

* [conch-hmac-sha2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bfc/bfcb719056a33e3e2f7d81e4492a2def90f5b851/8eb4ab3878e940ea75f5c2d5a646727ff2fde760.patch) (4988 bytes) - added by cjwatson on 2015-11-16 00:57:57Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8108 8108
type__enhancement enhancement
reporter__cjwatson cjwatson
priority__normal normal
milestone__None None
branch__branches_conch_ssh_hmac_sha2_8108 branches/conch-ssh-hmac-sha2-8108
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__None None
time__1447635290147012 1447635290147012
changetime__1447679376924578 1447679376924578
version__None None
owner__cjwatson cjwatson
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)<a name="note_2"></a>|@cjwatson removed owner|
|-|-|

This should do the job.  The only bit that was at all non-trivial is that SHA-512 has a different block size, so I needed to remove hardcoding of 64.

The test method names aren't ideal.  I would prefer test_hmac_sha2_512 etc. myself, but that doesn't seem to be permitted by Twisted's coding standards and my attempts to use underscores in test method names like that have failed review in the past.  Suggestions welcome if you also don't like this!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

(In [46228]) Branching to conch-ssh-hmac-sha2-8108.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to @cjwatson|
|-|-|

Thanks!

Changes looks good.

I also don't like the test names... but yes, this is the convention.
I will try to see if we can change this.

The current sha2-512 test is failing.... i think that this is due to a bad assertion.

I tried to fix the tests.

While looking a the code I found out that for `none` getMAC returns a simple tuple and the result no longer has the key attached.
I think that this is a bug and should be followd up in a separate ticket.

The new tests contain a lot of duplication.
I have tried to refactor this into a custom assertion call.

While working on this I have also update the client example to show the logs.

If you have time, please consider reviewing [#7715](https://github.com/twisted/twisted/issues/7715) so that we can also have an up to date sample ssh server.

Please check my latest changes and make sure they are OK.

Changes sent to buildbot.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)<a name="note_5"></a>|@cjwatson set owner to @adiroiban|
|-|-|

Thanks for fixing that up.  I swear the tests passed for me before I submitted this!  I must have made a mistake somewhere ...

Anyway, your test corrections look fine.  I think it would be more readable if you passed blockSize rather than blockPadSize to assertGetMAC, though, and had assertGetMAC calculate the required padding with (blockSize - digestSize).  Precomputing the padding makes it harder to do things like "git grep -w 128 twisted/conch" to look for places where block sizes are hardcoded.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @cjwatson|
|-|-|

I did not wanted to pass the blockSize to avoid duplicating the same logic used in the production code.

Why would you want to look for hardcoded block sizes in tests?

I think that in test, those values should be hardcoded or decoupled from the production code so that the checks are somehow more reliable.

Again, many thanks for the patch!

re-assigning before merge

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46231]) Merge conch-ssh-hmac-sha2-8108: Add support in twisted.conch.ssh for hmac-sha2-256 and hmac-sha2-512.

Author: cjwatson
Reviewer: adiroiban
Fixes: [#8108](https://github.com/twisted/twisted/issues/8108)
